File: programmers/43165.cpp
/*
- dfs, bfs 를 이용해 결과가 target이면 answer에 더하는형식
- 이번엔 c++을 bfs로 구현해보도록 합시다

- bfs <- queue를 이용해서 구현
*/
#include <string>
#include <vector>
#include <queue>

#include <iostream>
using namespace std;

int solution(vector<int> numbers, int target) { 
    int answer = 0;
    
    //넣을 숫자 인덱스
    int size = numbers.size()-1;
    queue<pair<int, int>> q;
    
    // 시작 세트 초기화
    q.push(make_pair(0, numbers[0]));
    q.push(make_pair(0, -numbers[0]));
    
    int count = 0;
    
    while(!q.empty()){
        count ++;
        pair<int, int> p = q.front();
        q.pop();
        
        if(p.first == size && p.second == target) {
            answer++;
        }
        
        if(p.first < size){
            int idx = ++p.first;
            q.push(make_pair(idx, p.second+numbers[idx]));
            q.push(make_pair(idx, p.second-numbers[idx]));  
        }
    }
    return answer;
}

File: programmers/43165.py
# dfs로 풀어보기
# target에 도달하면 끝
# 안 도달하고 더 넘기려면?! 어떻게 짜야하드라.,.,
# python은 call by reference 형식 노 노 . 배열같은건 전달되지만 값은 안된다.

answer = 0;
def dfs(idx, numbers, value, target):
    global answer;
    if(idx == len(numbers)):
        if(value==target):
            answer += 1
            return
    else:
        dfs(idx+1, numbers, value+numbers[idx], target)
        dfs(idx+1, numbers, value-numbers[idx], target)
    
def solution(numbers, target):
    
    dfs(0, numbers, 0, target)
    
    return answer